### PR TITLE
[SPARK-34592][WebUI] Mark indeterminate RDD in Web UI

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to mark indeterminate RDD in Web UI. 

### Why are the changes needed?

It is somehow hard to track which part is indeterminate in a graph of RDDs. In some cases we may need to track indeterminate RDDs. For example, indeterminate map stage fails and Spark is unable to fallback some parent stages. The developers are usually unable to easily identify indeterminate part from the complicated RDD computation. If Web UI can show up indeterminate RDD like cached RDD, it could be useful to track it.

### Does this PR introduce _any_ user-facing change?

Yes, there is a UI change for users.

### How was this patch tested?

Manual check with Web UI locally. Updated existing unit tests.

Screen Shot 2021-03-02 at 12 38 02 AM
Screen Shot 2021-03-05 at 9 27 14 AM

### CODE REVIEW

**Comment by mridulm** [comment]
Thanks, this is a really welcome change !

**Comment by mridulm** [comment]
The changes look good to me, but I am not very knowledgable about the UI area ... would be better for someone with UI knowhow to also look at it.

**Comment by holdenk** [comment]
nit: I think this introduces a user facing change, namely the UI change.

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -2013,7 +2013,7 @@ abstract class RDD[T: ClassTag](
    * Returns the deterministic level of this RDD's output. Please refer to [[DeterministicLevel]]
    * for the definition.
    *
-   * By default, an reliably checkpointed RDD, or RDD without parents(root RDD) is DETERMINATE. For
+   *l By default, an reliably checkpointed RDD, or RDD without parents(root RDD) is DETERMINATE. For
```
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala:2016]: nit: is the l intentional?

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 2]
```diff
@@ -2022,7 +2022,7 @@ abstract class RDD[T: ClassTag](
   // TODO: make it public so users can set deterministic level to their custom RDDs.
   // TODO: this can be per-partition. e.g. UnionRDD can have different deterministic level for
   // different partitions.
-  private[spark] final lazy val outputDeterministicLevel: DeterministicLevel.Value = {
+  private[spark] final def outputDeterministicLevel: DeterministicLevel.Value = {
```
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala:2025]: Should expose this given the TODO?

File: core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala [hunk 3]
```diff
@@ -1171,6 +1171,7 @@ private[spark] object JsonProtocolSuite extends Assertions {
       |          "Replication": 1
       |        },
       |        "Barrier" : false,
+      |        "DeterministicLevel" : "DETERMINATE",
```
Comment (holdenk) [inline core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:1174]: It might be good to add an INDETERMINATE test case to this file.

**Review by holdenk — COMMENTED** [review]
Thanks for working on this, I know it's currently WIP but shared some early thoughts

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 1 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/rdd/RDD.scala:2016]: It's a typo. :) Thanks for catching it.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 2 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/rdd/RDD.scala:2025]: I'd prefer not to expose it here. At least I don't hear a requirement to expose this.

File: core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala (hunk 3 above)
Comment (viirya) [reply core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:1174]: Okay. let me add a test case for it.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 2 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/rdd/RDD.scala:2025]: Sounds reasonable. Can we add a JIRA to track this and put the JIRA next to the TODO so we eventually come back and make a decision?

Comment (viirya) [reply core/src/main/scala/org/apache/spark/rdd/RDD.scala:2025]: Yea, sure. Let me add a JIRA here.

**Comment by viirya** [comment]
@mridulm @holdenk Thanks for comments and review. This is out of WIP and ready for review now.

**Comment by viirya** [comment]
also cc @dongjoon-hyun

File: core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala [hunk 4]
```diff
@@ -246,8 +254,14 @@ private[spark] object RDDOperationGraph extends Logging {
     } else {
       ""
     }
+    val outputDeterministicLevel = node.outputDeterministicLevel match {
+      case DeterministicLevel.DETERMINATE => ""
+      case DeterministicLevel.INDETERMINATE => " [indeterminate]"
+      case DeterministicLevel.UNORDERED => " [unordered]"
+    }
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala:261]: To be consistent with `[Barrier]` and `[Cached]`, t would be better to capitalize them.

**Comment by sarutak** [comment]
I've left one minor comment. Otherwise, from the point of view of UI, it LGTM.

**Review by holdenk — APPROVED** [review]
LGTM pending the consistency change suggested by sarutak

File: core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala (hunk 4 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/ui/scope/RDDOperationGraph.scala:261]: Thanks for the suggestion. Changed to capitalized ones.

**Comment by viirya** [comment]
Thanks @holdenk @sarutak for the review. I might open this for a few days and merge it if no more comments.

**Comment by maropu** [comment]
Looks a useful feature, @viirya !

**Review by sarutak — APPROVED** [review]

**Comment by holdenk** [comment]
I'm excited to see this in 3.2, I know when I'm helping users debug jobs having this information in the UI would be useful :)

**Review by dongjoon-hyun — COMMENTED** [review]
Looks useful. Thanks, @viirya .

cc @gengliangwang

File: core/src/main/scala/org/apache/spark/ui/UIUtils.scala [hunk 5]
```diff
@@ -506,6 +506,7 @@ private[spark] object UIUtils extends Logging {
         <a data-toggle="tooltip" title={if (forJob) ToolTips.JOB_DAG else ToolTips.STAGE_DAG}
            data-placement="top">
           DAG Visualization
+
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/ui/UIUtils.scala:509]: Nit: unnecessary change

**Comment by gengliangwang** [comment]
@viirya Thanks for the work! It would be great if we can further explain the "indeterminate" mark in Spark docs(e.g. in https://spark.apache.org/docs/3.1.1/web-ui.html). 
We can do it in this PR or another follow-up.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 6]
```diff
@@ -2019,14 +2022,14 @@ abstract class RDD[T: ClassTag](
    * candidate that is deterministic least. Please override [[getOutputDeterministicLevel]] to
    * provide custom logic of calculating output deterministic level.
    */
-  // TODO: make it public so users can set deterministic level to their custom RDDs.
+  // TODO(SPARK-34612): make it public so users can set deterministic level to their custom RDDs.
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala:2025]: Thank you for making it IDed TODO.

**Comment by viirya** [comment]
Thanks @gengliangwang! I will update the doc in a follow-up.

**Review by gengliangwang — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. And, +1 for the follow-up for the documentation.

**Comment by viirya** [comment]
Thank you @dongjoon-hyun @gengliangwang.

I think this receives enough approvals so I will merge it tomorrow morning.

**Comment by sarutak** [comment]
@viirya Now that you have done to capitalize the labels in the graph, could you update the captures in the description?

**Comment by viirya** [comment]
@sarutak Yes, I updated it. Thanks for reminding.

**Comment by viirya** [comment]
Thanks all! Merging to master!

**Comment by xuanyuanking** [comment]
Thanks for adding this UI changing. Very cool for reminding the indeterminate RDD/Stage!!
